File: 142-linked-list-cycle-ii/linked-list-cycle-ii.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        temp=head
        dict={}
        while(temp!=None):
            if(temp in dict):
                return temp
            dict[temp]=1
            temp=temp.next
        return None

File: 142-linked-list-cycle-ii/linked-list-cycle-ii.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        
        if(head==NULL){
            return NULL;
        }
        ListNode* slow=head,*fast=head;
        
        while(true){
            slow=slow->next;
            fast=fast->next;
            if(fast==NULL){
                return NULL;
            }
            fast=fast->next;
            if(fast==NULL){
                return NULL;
            }
            if(slow==fast){
                break;
            }
        }
        ListNode* ptr=head;
        while(slow!=ptr){
            ptr=ptr->next;
            slow=slow->next;
        }
        return ptr;
    }
};